readthedocs/proxito/views.py

```diff
@@ -0,0 +1,247 @@
+    if all([lang_slug is None, version_slug is None, filename == '',
+            not current_project.single_version]):
+        log.info('Proxito redirect: slug=%s', current_project.slug)
+        return redirect_project_slug(
+            request, project=current_project, subproject=None
+        )
+
+    if (lang_slug is None or version_slug is None) and not current_project.single_version:
+        log.info('Invalid URL for project with versions. url=%s', filename)
+        raise Http404('Invalid URL for project with versions')
+
+    # Handle single-version projects that have URLs like a real project
+    if current_project.single_version:
+        if lang_slug and version_slug:
+            filename = os.path.join(lang_slug, version_slug, filename)
+            lang_slug = version_slug = None
+
+    # Check to see if we need to serve a translation
+    if not lang_slug or lang_slug == current_project.language:
+        final_project = current_project
+    else:
+        final_project = get_object_or_404(
+            current_project.translations.all(), language=lang_slug
+        )
+
+    # ``final_project`` is now the actual project we want to serve docs on,
+    # accounting for:
+    # * Project
+    # * Subproject
+    # * Translations
+
+    # TODO: Redirects need to be refactored before we can turn them on
+    # They currently do 1 request per redirect that exists for the project
+    # path, http_status = final_project.redirects.get_redirect_path_with_status(
+    #     language=lang_slug, version_slug=version_slug, path=filename
+    # )
+
+    # Handle single version by grabbing the default version
+    if final_project.single_version:
+        version_slug = final_project.get_default_version()
+
+    # Don't do auth checks
+    # try:
+    #     Version.objects.public(user=request.user, project=final_project).get(slug=version_slug)
+    # except Version.DoesNotExist:
+    #     # Properly raise a 404 if the version doesn't exist (or is inactive) and
+    #     # a 401 if it does
+    #     if final_project.versions.filter(slug=version_slug, active=True).exists():
+    #         return _serve_401(request, final_project)
+    #     raise Http404('Version does not exist.')
+
+    storage_path = final_project.get_storage_path(
+        type_='html', version_slug=version_slug, include_file=False
+    )
+    path = f'{storage_path}/{filename}'
+
+    # Handle out backend storage not supporting directory indexes,
+    # so we need to append index.html when appropriate.
+    # TODO: We don't currently support `docs.example.com/en/latest/install`
+    # where the user actually wants `docs.example.com/en/latest/install/index.html`
+    # We would need to emulate nginx's try_files in order to handle this.
+    if path[-1] == '/':
+        path += 'index.html'
+
+    # Serve from the filesystem if using DEBUG or Testing
+    # Tests require this now since we don't want to check for the file existing in prod
```

Comment:
I'm not sure to follow this. Where this code is checking for file existing?

I understand that `PYTHON_MEDIA=True` is useful to run this locally and get docs served properly instead of just seeing the file served in the response. I'm not seeing why this is needed for testing, though.

Reply:
The django `serve` function will check for files existing. This comment is confusing though, I will rewrite it.